### PR TITLE
Delegated Tx Support

### PR DESCRIPTION
**Pull request by philipjames44**
* Adds in MsgExec, MsgRevoke, and MsgGrant authz functionality
* Adds in basic FeeGrant functionality
* Adds in AuthZ grant query functionality
* Implements toml reading & parsing for delegate tx workflows, e.g.:
```
[sender]
grantee_private_key_path = "/Users/phil/Desktop/peggyJV/ocular/test_dir/Zeus.pem"
grantee_account_number = 1
grantee_sequence_number = 0
granter_account = "somm1wdvjp39hg9ct2wkpslyfvnk0r7tup27jsd6j0a"
denom = "usomm"
gas_fee = 50000
gas_limit = 100000
timeout_height = 9001
memo = "Delegation memo"

[[transaction]]
name = "Dionysus"
destination_account = "somm1jlfkfh40pky9l2e38njcl58hampjha4852w6w4"
amount = 50

[[transaction]]
name = "Silenus"
destination_account = "somm1pk8l5tj50qqcuk2u0weu0smdmqpjvdg2k77krf"
amount = 500
```
* Adds in tx logging e.g. [1649286032871346000.txt](https://github.com/PeggyJV/ocular/files/8435595/1649286032871346000.txt)

* Remove MsgDelegate & MsgUndelegate functionality + testing due to signing bugs (didn't want to leave broken unused code, will bring back later in fixed state)

### CODE REVIEW

File: ocular/src/chain/client/tx.rs [hunk 1]
```diff
@@ -98,71 +164,10 @@ impl ChainClient {
...
-        // Build tx body.
-        let tx_body = match msg.to_any() {
-            Ok(msg) => tx::Body::new(vec![msg], &tx_metadata.memo, tx_metadata.timeout_height),
-            Err(err) => return Err(TxError::SerializationError(err.to_string())),
-        };
-
-        self.sign_and_send_msg(
-            delegator_account.public_key,
-            delegator_account.private_key,
-            amount,
```
Comment (philipjames44) [inline ocular/src/chain/client/tx.rs:163]: This is just a clean removal of delegate and undelegate msg types. Discovered some bugs with this while testing and decided to remove and bring back in fixed state later.

File: ocular/Cargo.toml [hunk 2]
```diff
@@ -14,7 +14,7 @@ readme = "README.md"
 assay = "0.1"
 bech32 = "0.6"
 bip32 = "0.3"
-cosmos-sdk-proto = { version = "0.10", features = ["grpc"] }
+cosmos-sdk-proto = { git = "https://github.com/cosmos/cosmos-rust", branch = "main", features = ["grpc"]}
```
Comment (philipjames44) [inline ocular/Cargo.toml:17]: We have to read from the latest version of main here as latest release does not have authz or feegrant modules exposed.

File: ocular/src/chain/client/authz.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        .await
+    }
+
+    // Basic fee allowance
+    pub async fn perform_basic_allowance_fee_grant(
+        &self,
+        granter: Account,
+        grantee: AccountId,
+        expiration: Option<prost_types::Timestamp>,
+        // TODO: Standardize below Coin type to common cosmrs coin type once FeeGrants get looped in.
```
Comment (philipjames44) [inline ocular/src/chain/client/authz.rs:133]: This todo is due to feegrants not being in cosmrs yet, so the coin type is the one defined by the cosmos-sdk

File: ocular/src/chain/client/automated_tx_handler.rs [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+    pub grant_gas_fee: u64,
+    pub grant_gas_limit: u64,
+    pub grant_timeout_height: u32,
+    pub grant_memo: &'a str,
+    // MsgExec data
+    pub exec_gas_fee: u64,
+    pub exec_gas_limit: u64,
+    pub exec_timeout_height: u32,
+    pub exec_memo: &'a str,
+}
```
Comment (philipjames44) [inline ocular/src/chain/client/automated_tx_handler.rs:54]: I'm open to alternative sender object design here. This is a bit verbose and depending on our needs/wants we can potentially simplify this a bit.

**Comment by zmanian** [comment]
My sense is the code doesn't permit sufficient flexibility in defining workflows.

I think a common case will be doing the grant from the somm binary to take advantage of the somm binaries ledger hardware wallet support.

Then using a toml file to execute sequence of transactions from the granted key.

When we move to more flexible workflows, it would be good if the code could also check via querying the authz module if a validate and un-expired grant for the message type exists before it starts executing.

**Review by zmanian — APPROVED** [review]
I think this looks good to merge

**Pull request merged**